#     mockqcdata.py: module providing mock Illumina data for testing
#     Copyright (C) University of Manchester 2012-2024 Peter Briggs
#
########################################################################

FASTQC_0_11_3 = {
    'fastqc_data.txt': """##FastQC	0.11.3
>>Basic Statistics	pass
#Measure	Value
Filename	%(fastq)s
File type	Conventional base calls
Encoding	Sanger / Illumina 1.9
Total Sequences	10
Sequences flagged as poor quality	0
Sequence length	250
%%GC	51
>>END_MODULE
>>Per base sequence quality	fail
#Base	Mean	Median	Lower Quartile	Upper Quartile	10th Percentile	90th Percentile
1	27.6	0.0	0.0	0.0	0.0	0.0
2	25.6	0.0	0.0	0.0	0.0	0.0
3	22.0	0.0	0.0	0.0	0.0	0.0
4	26.8	0.0	0.0	0.0	0.0	0.0
5	28.4	0.0	0.0	0.0	0.0	0.0
6	31.3	0.0	0.0	0.0	0.0	0.0
7	29.4	0.0	0.0	0.0	0.0	0.0
8	29.7	0.0	0.0	0.0	0.0	0.0
9	31.9	0.0	0.0	0.0	0.0	0.0
10-14	31.639999999999997	0.0	0.0	0.0	0.0	0.0
15-19	34.48	0.0	0.0	0.0	0.0	0.0
20-24	30.339999999999996	0.0	0.0	0.0	0.0	0.0
25-29	34.019999999999996	0.0	0.0	0.0	0.0	0.0
30-34	34.839999999999996	0.0	0.0	0.0	0.0	0.0
35-39	35.6	0.0	0.0	0.0	0.0	0.0
40-44	34.56	0.0	0.0	0.0	0.0	0.0
45-49	36.36	0.0	0.0	0.0	0.0	0.0
50-54	35.279999999999994	0.0	0.0	0.0	0.0	0.0
55-59	35.580000000000005	0.0	0.0	0.0	0.0	0.0
60-64	35.56	0.0	0.0	0.0	0.0	0.0
65-69	34.58	0.0	0.0	0.0	0.0	0.0
70-74	34.67999999999999	0.0	0.0	0.0	0.0	0.0
75-79	34.17999999999999	0.0	0.0	0.0	0.0	0.0
80-84	32.220000000000006	0.0	0.0	0.0	0.0	0.0
85-89	34.22	0.0	0.0	0.0	0.0	0.0
90-94	34.49999999999999	0.0	0.0	0.0	0.0	0.0
95-99	34.3	0.0	0.0	0.0	0.0	0.0
100-104	33.52	0.0	0.0	0.0	0.0	0.0
105-109	32.06	0.0	0.0	0.0	0.0	0.0
110-114	28.2	0.0	0.0	0.0	0.0	0.0
115-119	30.639999999999997	0.0	0.0	0.0	0.0	0.0
120-124	28.840000000000003	0.0	0.0	0.0	0.0	0.0
125-129	28.96	0.0	0.0	0.0	0.0	0.0
130-134	25.860000000000003	0.0	0.0	0.0	0.0	0.0
135-139	27.660000000000004	0.0	0.0	0.0	0.0	0.0
140-144	26.78	0.0	0.0	0.0	0.0	0.0
145-149	23.96	0.0	0.0	0.0	0.0	0.0
150-154	23.9	0.0	0.0	0.0	0.0	0.0
155-159	25.2	0.0	0.0	0.0	0.0	0.0
160-164	22.020000000000003	0.0	0.0	0.0	0.0	0.0
165-169	21.639999999999997	0.0	0.0	0.0	0.0	0.0
170-174	17.72	0.0	0.0	0.0	0.0	0.0
175-179	16.36	0.0	0.0	0.0	0.0	0.0
180-184	15.540000000000001	0.0	0.0	0.0	0.0	0.0
185-189	15.88	0.0	0.0	0.0	0.0	0.0
190-194	19.82	0.0	0.0	0.0	0.0	0.0
195-199	16.419999999999998	0.0	0.0	0.0	0.0	0.0
200-204	17.76	0.0	0.0	0.0	0.0	0.0
205-209	15.319999999999999	0.0	0.0	0.0	0.0	0.0
210-214	16.9	0.0	0.0	0.0	0.0	0.0
215-219	17.080000000000002	0.0	0.0	0.0	0.0	0.0
220-224	14.839999999999998	0.0	0.0	0.0	0.0	0.0
225-229	11.760000000000002	0.0	0.0	0.0	0.0	0.0
230-234	13.74	0.0	0.0	0.0	0.0	0.0
235-239	12.54	0.0	0.0	0.0	0.0	0.0
240-244	8.14	0.0	0.0	0.0	0.0	0.0
245-249	3.7399999999999998	0.0	0.0	0.0	0.0	0.0
250	4.5	0.0	0.0	0.0	0.0	0.0
>>END_MODULE
>>Per tile sequence quality	pass
#Tile	Base	Mean
1101	1	0.0
1101	2	0.0
1101	3	0.0
1101	4	0.0
1101	5	0.0
1101	6	0.0
1101	7	0.0
1101	8	0.0
1101	9	0.0
1101	10-14	0.0
1101	15-19	0.0
1101	20-24	0.0
1101	25-29	0.0
1101	30-34	0.0
1101	35-39	0.0
1101	40-44	0.0
1101	45-49	0.0
1101	50-54	0.0
1101	55-59	0.0
1101	60-64	0.0
1101	65-69	0.0
1101	70-74	0.0
1101	75-79	0.0
1101	80-84	0.0
1101	85-89	0.0
1101	90-94	0.0
1101	95-99	0.0
1101	100-104	0.0
1101	105-109	0.0
1101	110-114	0.0
1101	115-119	0.0
1101	120-124	0.0
1101	125-129	0.0
1101	130-134	0.0
1101	135-139	0.0
1101	140-144	0.0
1101	145-149	0.0
1101	150-154	0.0
1101	155-159	0.0
1101	160-164	0.0
1101	165-169	0.0
1101	170-174	0.0
1101	175-179	0.0
1101	180-184	0.0
1101	185-189	0.0
1101	190-194	0.0
1101	195-199	0.0
1101	200-204	0.0
1101	205-209	0.0
1101	210-214	0.0
1101	215-219	0.0
1101	220-224	0.0
1101	225-229	0.0
1101	230-234	0.0
1101	235-239	0.0
1101	240-244	0.0
1101	245-249	0.0
1101	250	0.0
>>END_MODULE
>>Per sequence quality scores	pass
#Quality	Count
12	1.0
13	0.0
14	0.0
15	0.0
16	1.0
17	0.0
18	1.0
19	0.0
20	0.0
21	0.0
22	1.0
23	0.0
24	0.0
25	0.0
26	0.0
27	1.0
28	0.0
29	1.0
30	1.0
31	1.0
32	0.0
33	2.0
>>END_MODULE
>>Per base sequence content	fail
#Base	G	A	T	C
1	40.0	30.0	10.0	20.0
2	20.0	30.0	40.0	10.0
3	20.0	40.0	40.0	0.0
4	10.0	20.0	20.0	50.0
5	0.0	60.0	30.0	10.0
6	40.0	20.0	10.0	30.0
7	10.0	20.0	40.0	30.0
8	10.0	20.0	30.0	40.0
9	40.0	10.0	0.0	50.0
10-14	26.0	22.0	20.0	32.0
15-19	24.0	46.0	12.0	18.0
20-24	21.73913043478261	26.08695652173913	21.73913043478261	30.434782608695656
25-29	20.0	34.0	18.0	28.000000000000004
30-34	22.0	14.000000000000002	28.000000000000004	36.0
35-39	30.0	34.0	18.0	18.0
40-44	22.0	22.0	30.0	26.0
45-49	28.000000000000004	24.0	24.0	24.0
50-54	34.0	20.0	20.0	26.0
55-59	24.0	24.0	12.0	40.0
60-64	12.0	28.000000000000004	32.0	28.000000000000004
65-69	28.000000000000004	16.0	34.0	22.0
70-74	22.0	26.0	24.0	28.000000000000004
75-79	32.0	16.0	20.0	32.0
80-84	24.0	22.0	22.0	32.0
85-89	14.000000000000002	24.0	34.0	28.000000000000004
90-94	28.000000000000004	28.000000000000004	28.000000000000004	16.0
95-99	28.000000000000004	26.0	10.0	36.0
100-104	20.0	16.0	34.0	30.0
105-109	18.0	18.0	24.0	40.0
110-114	30.0	20.0	18.0	32.0
115-119	18.0	28.000000000000004	30.0	24.0
120-124	26.0	28.000000000000004	18.0	28.000000000000004
125-129	14.000000000000002	26.0	22.0	38.0
130-134	32.0	30.0	22.0	16.0
135-139	22.0	22.0	32.0	24.0
140-144	20.0	36.0	16.0	28.000000000000004
145-149	32.0	18.0	28.000000000000004	22.0
150-154	26.0	26.0	28.000000000000004	20.0
155-159	16.0	22.0	24.0	38.0
160-164	26.0	20.0	36.0	18.0
165-169	30.0	24.0	18.0	28.000000000000004
170-174	28.000000000000004	16.0	24.0	32.0
175-179	24.0	34.0	14.000000000000002	28.000000000000004
180-184	24.0	24.0	20.0	32.0
185-189	22.0	24.0	24.0	30.0
190-194	16.0	14.000000000000002	36.0	34.0
195-199	22.0	30.0	20.0	28.000000000000004
200-204	24.0	30.0	26.0	20.0
205-209	24.0	22.0	26.0	28.000000000000004
210-214	20.0	22.0	26.0	32.0
215-219	18.0	24.0	28.000000000000004	30.0
220-224	18.0	24.0	22.0	36.0
225-229	26.0	24.0	20.0	30.0
230-234	20.0	30.0	18.0	32.0
235-239	28.000000000000004	16.0	18.0	38.0
240-244	28.000000000000004	32.0	22.0	18.0
245-249	22.0	34.0	12.0	32.0
250	30.0	40.0	10.0	20.0
>>END_MODULE
>>Per sequence GC content	fail
#GC Content	Count
0	0.0
1	0.0
2	0.0
3	0.0
4	0.0
5	0.0
6	0.0
7	0.0
8	0.0
9	0.0
10	0.0
11	0.0
12	0.0
13	0.0
14	0.0
15	0.0
16	0.0
17	0.0
18	0.0
19	0.0
20	0.0
21	0.0
22	0.0
23	0.0
24	0.0
25	0.0
26	0.0
27	0.0
28	0.0
29	0.0
30	0.0
31	0.0
32	0.0
33	0.0
34	0.0
35	0.0
36	0.0
37	0.0
38	0.0
39	0.3333333333333333
40	0.3333333333333333
41	0.0
42	0.0
43	0.0
44	0.0
45	0.0
46	0.0
47	0.3333333333333333
48	0.3333333333333333
49	0.6666666666666666
50	0.0
51	1.0
52	0.3333333333333333
53	0.0
54	0.0
55	0.0
56	0.0
57	0.0
58	0.3333333333333333
59	0.6666666666666666
60	0.0
61	0.0
62	0.3333333333333333
63	0.0
64	0.0
65	0.0
66	0.0
67	0.0
68	0.0
69	0.0
70	0.0
71	0.0
72	0.0
73	0.0
74	0.0
75	0.0
76	0.0
77	0.0
78	0.0
79	0.0
80	0.0
81	0.0
82	0.0
83	0.0
84	0.0
85	0.0
86	0.0
87	0.0
88	0.0
89	0.0
90	0.0
91	0.0
92	0.0
93	0.0
94	0.0
95	0.0
96	0.0
97	0.0
98	0.0
99	0.0
100	0.0
>>END_MODULE
>>Per base N content	warn
#Base	N-Count
1	0.0
2	0.0
3	0.0
4	0.0
5	0.0
6	0.0
7	0.0
8	0.0
9	0.0
10-14	0.0
15-19	0.0
20-24	8.0
25-29	0.0
30-34	0.0
35-39	0.0
40-44	0.0
45-49	0.0
50-54	0.0
55-59	0.0
60-64	0.0
65-69	0.0
70-74	0.0
75-79	0.0
80-84	0.0
85-89	0.0
90-94	0.0
95-99	0.0
100-104	0.0
105-109	0.0
110-114	0.0
115-119	0.0
120-124	0.0
125-129	0.0
130-134	0.0
135-139	0.0
140-144	0.0
145-149	0.0
150-154	0.0
155-159	0.0
160-164	0.0
165-169	0.0
170-174	0.0
175-179	0.0
180-184	0.0
185-189	0.0
190-194	0.0
195-199	0.0
200-204	0.0
205-209	0.0
210-214	0.0
215-219	0.0
220-224	0.0
225-229	0.0
230-234	0.0
235-239	0.0
240-244	0.0
245-249	0.0
250	0.0
>>END_MODULE
>>Sequence Length Distribution	pass
#Length	Count
250	10.0
>>END_MODULE
>>Sequence Duplication Levels	pass
#Total Deduplicated Percentage	100.0
#Duplication Level	Percentage of deduplicated	Percentage of total
1	100.0	100.0
2	0.0	0.0
3	0.0	0.0
4	0.0	0.0
5	0.0	0.0
6	0.0	0.0
7	0.0	0.0
8	0.0	0.0
9	0.0	0.0
>10	0.0	0.0
>50	0.0	0.0
>100	0.0	0.0
>500	0.0	0.0
>1k	0.0	0.0
>5k	0.0	0.0
>10k+	0.0	0.0
>>END_MODULE
>>Overrepresented sequences	fail
#Sequence	Count	Percentage	Possible Source
GGTATCCCCCGGCAGTGAGGATGGAGCCATGGTCTGCATCATACTCACCG	1	10.0	No Hit
GAGCAGTCGGGCTCAGCGCTNTGCAAATTCTAGTTAGAAACTCACAGTTC	1	10.0	No Hit
AAAATAATCCTAAAAAATAACCTCTATGCCGCCGAACGCTCCGCCTCTAT	1	10.0	No Hit
GTAGTATTCTCATATCACAAGTCCCCAAACTGCATAAGGTGTGGAGTGGA	1	10.0	No Hit
ATATATTCATCCGCCATTATNAGAGTCCGATTACTTTAGAACAGTGCCGC	1	10.0	No Hit
CATCACTACCGCTCAGGAATNTGACGGCAGTCTTAGCGGCGCTCTAGTGC	1	10.0	No Hit
AGATAGCCGAAGATAAAGAGNTCATAACCGTAAAGGCCAGAGACGAGAAC	1	10.0	No Hit
GTGCAGGGGGTGTGGTCAATCCACACTGTTGCTGAGGTGATTGGGTCTCC	1	10.0	No Hit
TCTCAGATGAGCATGCAGCAGCCCAGACTCGCCCCACGCAGTTTGCCAAC	1	10.0	No Hit
CTTCCCCACGGCCCAGACACAAGAGACGACCTCCATAAATCTTTTAGAGG	1	10.0	No Hit
>>END_MODULE
>>Adapter Content	fail
#Position	Illumina Universal Adapter	Illumina Small RNA Adapter	Nextera Transposase Sequence	SOLID Small RNA Adapter
1	0.0	0.0	0.0	0.0
2	0.0	0.0	0.0	0.0
3	0.0	0.0	0.0	0.0
4	0.0	0.0	0.0	0.0
5	0.0	0.0	0.0	0.0
6	0.0	0.0	0.0	0.0
7	0.0	0.0	0.0	0.0
8	0.0	0.0	0.0	0.0
9	0.0	0.0	0.0	0.0
10-14	0.0	0.0	0.0	0.0
15-19	0.0	0.0	0.0	0.0
20-24	0.0	0.0	0.0	0.0
25-29	0.0	0.0	0.0	0.0
30-34	0.0	0.0	0.0	0.0
35-39	0.0	0.0	0.0	0.0
40-44	0.0	0.0	0.0	0.0
45-49	0.0	0.0	0.0	0.0
50-54	0.0	0.0	0.0	0.0
55-59	0.0	0.0	0.0	0.0
60-64	0.0	0.0	0.0	0.0
65-69	0.0	0.0	0.0	0.0
70-74	0.0	0.0	0.0	0.0
75-79	0.0	0.0	0.0	0.0
80-84	0.0	0.0	0.0	0.0
85-89	0.0	0.0	0.0	0.0
90-94	0.0	0.0	0.0	0.0
95-99	0.0	0.0	0.0	0.0
100-104	0.0	0.0	0.0	0.0
105-109	0.0	0.0	2.0	0.0
110-114	0.0	0.0	10.0	0.0
115-119	0.0	0.0	10.0	0.0
120-124	0.0	0.0	10.0	0.0
125-129	0.0	0.0	10.0	0.0
130-134	0.0	0.0	10.0	0.0
135-139	0.0	0.0	10.0	0.0
140-144	0.0	0.0	16.0	0.0
145-149	0.0	0.0	20.0	0.0
150-154	0.0	0.0	20.0	0.0
155-159	0.0	0.0	20.0	0.0
160-164	0.0	0.0	20.0	0.0
165-169	0.0	0.0	20.0	0.0
170-174	0.0	0.0	20.0	0.0
175-179	0.0	0.0	20.0	0.0
180-184	0.0	0.0	20.0	0.0
185-189	0.0	0.0	20.0	0.0
190-194	0.0	0.0	20.0	0.0
195-199	0.0	0.0	20.0	0.0
200-204	0.0	0.0	20.0	0.0
205-209	0.0	0.0	20.0	0.0
210-214	0.0	0.0	22.0	0.0
215-219	0.0	0.0	30.0	0.0
220-224	0.0	0.0	30.0	0.0
225-229	0.0	0.0	30.0	0.0
230-234	0.0	0.0	30.0	0.0
235-238	0.0	0.0	30.0	0.0
>>END_MODULE
>>Kmer Content	pass
>>END_MODULE
##FastQC	0.11.3
>>Basic Statistics	pass
#Measure	Value
Filename	Illumina_SG_R1.fastq
File type	Conventional base calls
Encoding	Sanger / Illumina 1.9
Total Sequences	10
Sequences flagged as poor quality	0
Sequence length	250
%%GC	51
>>END_MODULE
>>Per base sequence quality	fail
#Base	Mean	Median	Lower Quartile	Upper Quartile	10th Percentile	90th Percentile
1	27.6	0.0	0.0	0.0	0.0	0.0
2	25.6	0.0	0.0	0.0	0.0	0.0
3	22.0	0.0	0.0	0.0	0.0	0.0
4	26.8	0.0	0.0	0.0	0.0	0.0
5	28.4	0.0	0.0	0.0	0.0	0.0
6	31.3	0.0	0.0	0.0	0.0	0.0
7	29.4	0.0	0.0	0.0	0.0	0.0
8	29.7	0.0	0.0	0.0	0.0	0.0
9	31.9	0.0	0.0	0.0	0.0	0.0
10-14	31.639999999999997	0.0	0.0	0.0	0.0	0.0
15-19	34.48	0.0	0.0	0.0	0.0	0.0
20-24	30.339999999999996	0.0	0.0	0.0	0.0	0.0
25-29	34.019999999999996	0.0	0.0	0.0	0.0	0.0
30-34	34.839999999999996	0.0	0.0	0.0	0.0	0.0
35-39	35.6	0.0	0.0	0.0	0.0	0.0
40-44	34.56	0.0	0.0	0.0	0.0	0.0
45-49	36.36	0.0	0.0	0.0	0.0	0.0
50-54	35.279999999999994	0.0	0.0	0.0	0.0	0.0
55-59	35.580000000000005	0.0	0.0	0.0	0.0	0.0
60-64	35.56	0.0	0.0	0.0	0.0	0.0
65-69	34.58	0.0	0.0	0.0	0.0	0.0
70-74	34.67999999999999	0.0	0.0	0.0	0.0	0.0
75-79	34.17999999999999	0.0	0.0	0.0	0.0	0.0
80-84	32.220000000000006	0.0	0.0	0.0	0.0	0.0
85-89	34.22	0.0	0.0	0.0	0.0	0.0
90-94	34.49999999999999	0.0	0.0	0.0	0.0	0.0
95-99	34.3	0.0	0.0	0.0	0.0	0.0
100-104	33.52	0.0	0.0	0.0	0.0	0.0
105-109	32.06	0.0	0.0	0.0	0.0	0.0
110-114	28.2	0.0	0.0	0.0	0.0	0.0
115-119	30.639999999999997	0.0	0.0	0.0	0.0	0.0
120-124	28.840000000000003	0.0	0.0	0.0	0.0	0.0
125-129	28.96	0.0	0.0	0.0	0.0	0.0
130-134	25.860000000000003	0.0	0.0	0.0	0.0	0.0
135-139	27.660000000000004	0.0	0.0	0.0	0.0	0.0
140-144	26.78	0.0	0.0	0.0	0.0	0.0
145-149	23.96	0.0	0.0	0.0	0.0	0.0
150-154	23.9	0.0	0.0	0.0	0.0	0.0
155-159	25.2	0.0	0.0	0.0	0.0	0.0
160-164	22.020000000000003	0.0	0.0	0.0	0.0	0.0
165-169	21.639999999999997	0.0	0.0	0.0	0.0	0.0
170-174	17.72	0.0	0.0	0.0	0.0	0.0
175-179	16.36	0.0	0.0	0.0	0.0	0.0
180-184	15.540000000000001	0.0	0.0	0.0	0.0	0.0
185-189	15.88	0.0	0.0	0.0	0.0	0.0
190-194	19.82	0.0	0.0	0.0	0.0	0.0
195-199	16.419999999999998	0.0	0.0	0.0	0.0	0.0
200-204	17.76	0.0	0.0	0.0	0.0	0.0
205-209	15.319999999999999	0.0	0.0	0.0	0.0	0.0
210-214	16.9	0.0	0.0	0.0	0.0	0.0
215-219	17.080000000000002	0.0	0.0	0.0	0.0	0.0
220-224	14.839999999999998	0.0	0.0	0.0	0.0	0.0
225-229	11.760000000000002	0.0	0.0	0.0	0.0	0.0
230-234	13.74	0.0	0.0	0.0	0.0	0.0
235-239	12.54	0.0	0.0	0.0	0.0	0.0
240-244	8.14	0.0	0.0	0.0	0.0	0.0
245-249	3.7399999999999998	0.0	0.0	0.0	0.0	0.0
250	4.5	0.0	0.0	0.0	0.0	0.0
>>END_MODULE
>>Per tile sequence quality	pass
#Tile	Base	Mean
1101	1	0.0
1101	2	0.0
1101	3	0.0
1101	4	0.0
1101	5	0.0
1101	6	0.0
1101	7	0.0
1101	8	0.0
1101	9	0.0
1101	10-14	0.0
1101	15-19	0.0
1101	20-24	0.0
1101	25-29	0.0
1101	30-34	0.0
1101	35-39	0.0
1101	40-44	0.0
1101	45-49	0.0
1101	50-54	0.0
1101	55-59	0.0
1101	60-64	0.0
1101	65-69	0.0
1101	70-74	0.0
1101	75-79	0.0
1101	80-84	0.0
1101	85-89	0.0
1101	90-94	0.0
1101	95-99	0.0
1101	100-104	0.0
1101	105-109	0.0
1101	110-114	0.0
1101	115-119	0.0
1101	120-124	0.0
1101	125-129	0.0
1101	130-134	0.0
1101	135-139	0.0
1101	140-144	0.0
1101	145-149	0.0
1101	150-154	0.0
1101	155-159	0.0
1101	160-164	0.0
1101	165-169	0.0
1101	170-174	0.0
1101	175-179	0.0
1101	180-184	0.0
1101	185-189	0.0
1101	190-194	0.0
1101	195-199	0.0
1101	200-204	0.0
1101	205-209	0.0
1101	210-214	0.0
1101	215-219	0.0
1101	220-224	0.0
1101	225-229	0.0
1101	230-234	0.0
1101	235-239	0.0
1101	240-244	0.0
1101	245-249	0.0
1101	250	0.0
>>END_MODULE
>>Per sequence quality scores	pass
#Quality	Count
12	1.0
13	0.0
14	0.0
15	0.0
16	1.0
17	0.0
18	1.0
19	0.0
20	0.0
21	0.0
22	1.0
23	0.0
24	0.0
25	0.0
26	0.0
27	1.0
28	0.0
29	1.0
30	1.0
31	1.0
32	0.0
33	2.0
>>END_MODULE
>>Per base sequence content	fail
#Base	G	A	T	C
1	40.0	30.0	10.0	20.0
2	20.0	30.0	40.0	10.0
3	20.0	40.0	40.0	0.0
4	10.0	20.0	20.0	50.0
5	0.0	60.0	30.0	10.0
6	40.0	20.0	10.0	30.0
7	10.0	20.0	40.0	30.0
8	10.0	20.0	30.0	40.0
9	40.0	10.0	0.0	50.0
10-14	26.0	22.0	20.0	32.0
15-19	24.0	46.0	12.0	18.0
20-24	21.73913043478261	26.08695652173913	21.73913043478261	30.434782608695656
25-29	20.0	34.0	18.0	28.000000000000004
30-34	22.0	14.000000000000002	28.000000000000004	36.0
35-39	30.0	34.0	18.0	18.0
40-44	22.0	22.0	30.0	26.0
45-49	28.000000000000004	24.0	24.0	24.0
50-54	34.0	20.0	20.0	26.0
55-59	24.0	24.0	12.0	40.0
60-64	12.0	28.000000000000004	32.0	28.000000000000004
65-69	28.000000000000004	16.0	34.0	22.0
70-74	22.0	26.0	24.0	28.000000000000004
75-79	32.0	16.0	20.0	32.0
80-84	24.0	22.0	22.0	32.0
85-89	14.000000000000002	24.0	34.0	28.000000000000004
90-94	28.000000000000004	28.000000000000004	28.000000000000004	16.0
95-99	28.000000000000004	26.0	10.0	36.0
100-104	20.0	16.0	34.0	30.0
105-109	18.0	18.0	24.0	40.0
110-114	30.0	20.0	18.0	32.0
115-119	18.0	28.000000000000004	30.0	24.0
120-124	26.0	28.000000000000004	18.0	28.000000000000004
125-129	14.000000000000002	26.0	22.0	38.0
130-134	32.0	30.0	22.0	16.0
135-139	22.0	22.0	32.0	24.0
140-144	20.0	36.0	16.0	28.000000000000004
145-149	32.0	18.0	28.000000000000004	22.0
150-154	26.0	26.0	28.000000000000004	20.0
155-159	16.0	22.0	24.0	38.0
160-164	26.0	20.0	36.0	18.0
165-169	30.0	24.0	18.0	28.000000000000004
170-174	28.000000000000004	16.0	24.0	32.0
175-179	24.0	34.0	14.000000000000002	28.000000000000004
180-184	24.0	24.0	20.0	32.0
185-189	22.0	24.0	24.0	30.0
190-194	16.0	14.000000000000002	36.0	34.0
195-199	22.0	30.0	20.0	28.000000000000004
200-204	24.0	30.0	26.0	20.0
205-209	24.0	22.0	26.0	28.000000000000004
210-214	20.0	22.0	26.0	32.0
215-219	18.0	24.0	28.000000000000004	30.0
220-224	18.0	24.0	22.0	36.0
225-229	26.0	24.0	20.0	30.0
230-234	20.0	30.0	18.0	32.0
235-239	28.000000000000004	16.0	18.0	38.0
240-244	28.000000000000004	32.0	22.0	18.0
245-249	22.0	34.0	12.0	32.0
250	30.0	40.0	10.0	20.0
>>END_MODULE
>>Per sequence GC content	fail
#GC Content	Count
0	0.0
1	0.0
2	0.0
3	0.0
4	0.0
5	0.0
6	0.0
7	0.0
8	0.0
9	0.0
10	0.0
11	0.0
12	0.0
13	0.0
14	0.0
15	0.0
16	0.0
17	0.0
18	0.0
19	0.0
20	0.0
21	0.0
22	0.0
23	0.0
24	0.0
25	0.0
26	0.0
27	0.0
28	0.0
29	0.0
30	0.0
31	0.0
32	0.0
33	0.0
34	0.0
35	0.0
36	0.0
37	0.0
38	0.0
39	0.3333333333333333
40	0.3333333333333333
41	0.0
42	0.0
43	0.0
44	0.0
45	0.0
46	0.0
47	0.3333333333333333
48	0.3333333333333333
49	0.6666666666666666
50	0.0
51	1.0
52	0.3333333333333333
53	0.0
54	0.0
55	0.0
56	0.0
57	0.0
58	0.3333333333333333
59	0.6666666666666666
60	0.0
61	0.0
62	0.3333333333333333
63	0.0
64	0.0
65	0.0
66	0.0
67	0.0
68	0.0
69	0.0
70	0.0
71	0.0
72	0.0
73	0.0
74	0.0
75	0.0
76	0.0
77	0.0
78	0.0
79	0.0
80	0.0
81	0.0
82	0.0
83	0.0
84	0.0
85	0.0
86	0.0
87	0.0
88	0.0
89	0.0
90	0.0
91	0.0
92	0.0
93	0.0
94	0.0
95	0.0
96	0.0
97	0.0
98	0.0
99	0.0
100	0.0
>>END_MODULE
>>Per base N content	warn
#Base	N-Count
1	0.0
2	0.0
3	0.0
4	0.0
5	0.0
6	0.0
7	0.0
8	0.0
9	0.0
10-14	0.0
15-19	0.0
20-24	8.0
25-29	0.0
30-34	0.0
35-39	0.0
40-44	0.0
45-49	0.0
50-54	0.0
55-59	0.0
60-64	0.0
65-69	0.0
70-74	0.0
75-79	0.0
80-84	0.0
85-89	0.0
90-94	0.0
95-99	0.0
100-104	0.0
105-109	0.0
110-114	0.0
115-119	0.0
120-124	0.0
125-129	0.0
130-134	0.0
135-139	0.0
140-144	0.0
145-149	0.0
150-154	0.0
155-159	0.0
160-164	0.0
165-169	0.0
170-174	0.0
175-179	0.0
180-184	0.0
185-189	0.0
190-194	0.0
195-199	0.0
200-204	0.0
205-209	0.0
210-214	0.0
215-219	0.0
220-224	0.0
225-229	0.0
230-234	0.0
235-239	0.0
240-244	0.0
245-249	0.0
250	0.0
>>END_MODULE
>>Sequence Length Distribution	pass
#Length	Count
250	10.0
>>END_MODULE
>>Sequence Duplication Levels	pass
#Total Deduplicated Percentage	100.0
#Duplication Level	Percentage of deduplicated	Percentage of total
1	100.0	100.0
2	0.0	0.0
3	0.0	0.0
4	0.0	0.0
5	0.0	0.0
6	0.0	0.0
7	0.0	0.0
8	0.0	0.0
9	0.0	0.0
>10	0.0	0.0
>50	0.0	0.0
>100	0.0	0.0
>500	0.0	0.0
>1k	0.0	0.0
>5k	0.0	0.0
>10k+	0.0	0.0
>>END_MODULE
>>Overrepresented sequences	fail
#Sequence	Count	Percentage	Possible Source
GGTATCCCCCGGCAGTGAGGATGGAGCCATGGTCTGCATCATACTCACCG	1	10.0	No Hit
GAGCAGTCGGGCTCAGCGCTNTGCAAATTCTAGTTAGAAACTCACAGTTC	1	10.0	No Hit
AAAATAATCCTAAAAAATAACCTCTATGCCGCCGAACGCTCCGCCTCTAT	1	10.0	No Hit
GTAGTATTCTCATATCACAAGTCCCCAAACTGCATAAGGTGTGGAGTGGA	1	10.0	No Hit
ATATATTCATCCGCCATTATNAGAGTCCGATTACTTTAGAACAGTGCCGC	1	10.0	No Hit
CATCACTACCGCTCAGGAATNTGACGGCAGTCTTAGCGGCGCTCTAGTGC	1	10.0	No Hit
AGATAGCCGAAGATAAAGAGNTCATAACCGTAAAGGCCAGAGACGAGAAC	1	10.0	No Hit
GTGCAGGGGGTGTGGTCAATCCACACTGTTGCTGAGGTGATTGGGTCTCC	1	10.0	No Hit
TCTCAGATGAGCATGCAGCAGCCCAGACTCGCCCCACGCAGTTTGCCAAC	1	10.0	No Hit
CTTCCCCACGGCCCAGACACAAGAGACGACCTCCATAAATCTTTTAGAGG	1	10.0	No Hit
>>END_MODULE
>>Adapter Content	fail
#Position	Illumina Universal Adapter	Illumina Small RNA Adapter	Nextera Transposase Sequence	SOLID Small RNA Adapter
1	0.0	0.0	0.0	0.0
2	0.0	0.0	0.0	0.0
3	0.0	0.0	0.0	0.0
4	0.0	0.0	0.0	0.0
5	0.0	0.0	0.0	0.0
6	0.0	0.0	0.0	0.0
7	0.0	0.0	0.0	0.0
8	0.0	0.0	0.0	0.0
9	0.0	0.0	0.0	0.0
10-14	0.0	0.0	0.0	0.0
15-19	0.0	0.0	0.0	0.0
20-24	0.0	0.0	0.0	0.0
25-29	0.0	0.0	0.0	0.0
30-34	0.0	0.0	0.0	0.0
35-39	0.0	0.0	0.0	0.0
40-44	0.0	0.0	0.0	0.0
45-49	0.0	0.0	0.0	0.0
50-54	0.0	0.0	0.0	0.0
55-59	0.0	0.0	0.0	0.0
60-64	0.0	0.0	0.0	0.0
65-69	0.0	0.0	0.0	0.0
70-74	0.0	0.0	0.0	0.0
75-79	0.0	0.0	0.0	0.0
80-84	0.0	0.0	0.0	0.0
85-89	0.0	0.0	0.0	0.0
90-94	0.0	0.0	0.0	0.0
95-99	0.0	0.0	0.0	0.0
100-104	0.0	0.0	0.0	0.0
105-109	0.0	0.0	2.0	0.0
110-114	0.0	0.0	10.0	0.0
115-119	0.0	0.0	10.0	0.0
120-124	0.0	0.0	10.0	0.0
125-129	0.0	0.0	10.0	0.0
130-134	0.0	0.0	10.0	0.0
135-139	0.0	0.0	10.0	0.0
140-144	0.0	0.0	16.0	0.0
145-149	0.0	0.0	20.0	0.0
150-154	0.0	0.0	20.0	0.0
155-159	0.0	0.0	20.0	0.0
160-164	0.0	0.0	20.0	0.0
165-169	0.0	0.0	20.0	0.0
170-174	0.0	0.0	20.0	0.0
175-179	0.0	0.0	20.0	0.0
180-184	0.0	0.0	20.0	0.0
185-189	0.0	0.0	20.0	0.0
190-194	0.0	0.0	20.0	0.0
195-199	0.0	0.0	20.0	0.0
200-204	0.0	0.0	20.0	0.0
205-209	0.0	0.0	20.0	0.0
210-214	0.0	0.0	22.0	0.0
215-219	0.0	0.0	30.0	0.0
220-224	0.0	0.0	30.0	0.0
225-229	0.0	0.0	30.0	0.0
230-234	0.0	0.0	30.0	0.0
235-238	0.0	0.0	30.0	0.0
>>END_MODULE
>>Kmer Content	pass
>>END_MODULE
""",
}

FASTQC_0_12_1 = {
    'fastqc_data.txt': """##FastQC	0.12.1
>>Basic Statistics	pass
#Measure	Value
Filename	%(fastq)s
File type	Conventional base calls
Encoding	Sanger / Illumina 1.9
Total Sequences	379988
Total Bases	28.6 Mbp
Sequences flagged as poor quality	0
Sequence length	35-76
%%GC	58
>>END_MODULE
>>Per base sequence quality	pass
#Base	Mean	Median	Lower Quartile	Upper Quartile	10th Percentile	90th Percentile
1	34.265750497384126	37.0	32.0	37.0	32.0	37.0
2	34.7533895807236	37.0	32.0	37.0	32.0	37.0
3	35.59621093297683	37.0	37.0	37.0	32.0	37.0
4	36.022850721601735	37.0	37.0	37.0	37.0	37.0
5	36.157636556943906	37.0	37.0	37.0	37.0	37.0
6	36.21441466572629	37.0	37.0	37.0	37.0	37.0
7	36.202543238207525	37.0	37.0	37.0	37.0	37.0
8	36.23282840510753	37.0	37.0	37.0	37.0	37.0
9	36.22757560765077	37.0	37.0	37.0	37.0	37.0
10	36.197772034906365	37.0	37.0	37.0	37.0	37.0
11	36.28089834415824	37.0	37.0	37.0	37.0	37.0
12	36.25795025106056	37.0	37.0	37.0	37.0	37.0
13	36.24621566996853	37.0	37.0	37.0	37.0	37.0
14	36.2271913849911	37.0	37.0	37.0	37.0	37.0
15	36.23392054485931	37.0	37.0	37.0	37.0	37.0
16	36.192695558807124	37.0	37.0	37.0	37.0	37.0
17	36.21524363927282	37.0	37.0	37.0	37.0	37.0
18	36.181968904281185	37.0	37.0	37.0	37.0	37.0
19	36.200869501142144	37.0	37.0	37.0	37.0	37.0
20	36.206522311230884	37.0	37.0	37.0	37.0	37.0
21	36.18904807520237	37.0	37.0	37.0	37.0	37.0
22	36.16506047559397	37.0	37.0	37.0	37.0	37.0
23	36.162520921713316	37.0	37.0	37.0	37.0	37.0
24	36.139520195374594	37.0	37.0	37.0	37.0	37.0
25	36.12896723054412	37.0	37.0	37.0	37.0	37.0
26	36.07914460456646	37.0	37.0	37.0	37.0	37.0
27	36.06772582292072	37.0	37.0	37.0	37.0	37.0
28	36.04159078707749	37.0	37.0	37.0	37.0	37.0
29	36.04337768561113	37.0	37.0	37.0	37.0	37.0
30	36.04325399749466	37.0	37.0	37.0	37.0	37.0
31	36.000600018947964	37.0	37.0	37.0	37.0	37.0
32	36.015558386054295	37.0	37.0	37.0	37.0	37.0
33	35.958885543754015	37.0	37.0	37.0	37.0	37.0
34	35.99739465456804	37.0	37.0	37.0	37.0	37.0
35	35.95662757771298	37.0	37.0	37.0	37.0	37.0
36	35.959577193093416	37.0	37.0	37.0	37.0	37.0
37	35.93002031872195	37.0	37.0	37.0	37.0	37.0
38	35.962544210587346	37.0	37.0	37.0	37.0	37.0
39	35.96932086839865	37.0	37.0	37.0	37.0	37.0
40	35.945413091896306	37.0	37.0	37.0	37.0	37.0
41	35.89877822150172	37.0	37.0	37.0	37.0	37.0
42	35.87243757381474	37.0	37.0	37.0	37.0	37.0
43	35.865439272738776	37.0	37.0	37.0	37.0	37.0
44	35.85471741893296	37.0	37.0	37.0	37.0	37.0
45	35.8450363643076	37.0	37.0	37.0	37.0	37.0
46	35.81072869689058	37.0	37.0	37.0	37.0	37.0
47	35.830276318288284	37.0	37.0	37.0	37.0	37.0
48	35.82931088606393	37.0	37.0	37.0	37.0	37.0
49	35.7986607814219	37.0	37.0	37.0	37.0	37.0
50	35.766490831316816	37.0	37.0	37.0	37.0	37.0
51	35.73521454645273	37.0	37.0	37.0	37.0	37.0
52	35.76437214042258	37.0	37.0	37.0	37.0	37.0
53	35.748663708868115	37.0	37.0	37.0	37.0	37.0
54	35.73334795082184	37.0	37.0	37.0	37.0	37.0
55	35.74270167992833	37.0	37.0	37.0	37.0	37.0
56	35.764572147628826	37.0	37.0	37.0	37.0	37.0
57	35.749015552811066	37.0	37.0	37.0	37.0	37.0
58	35.76422824456362	37.0	37.0	37.0	37.0	37.0
59	35.75217586955371	37.0	37.0	37.0	37.0	37.0
60	35.664236341004795	37.0	37.0	37.0	37.0	37.0
61	35.693948282577765	37.0	37.0	37.0	37.0	37.0
62	35.70856720161412	37.0	37.0	37.0	37.0	37.0
63	35.63793702721993	37.0	37.0	37.0	37.0	37.0
64	35.63855386494596	37.0	37.0	37.0	37.0	37.0
65	35.62965049833887	37.0	37.0	37.0	37.0	37.0
66	35.5008073375948	37.0	37.0	37.0	32.0	37.0
67	35.56156293683019	37.0	37.0	37.0	37.0	37.0
68	35.56868823096688	37.0	37.0	37.0	37.0	37.0
69	35.539922939814076	37.0	37.0	37.0	32.0	37.0
70	35.56268633590152	37.0	37.0	37.0	32.0	37.0
71	35.46554044472746	37.0	37.0	37.0	32.0	37.0
72	35.444957783662396	37.0	37.0	37.0	32.0	37.0
73	35.51330281906549	37.0	37.0	37.0	37.0	37.0
74	35.46663348247984	37.0	37.0	37.0	32.0	37.0
75	35.38018079135356	37.0	37.0	37.0	32.0	37.0
76	34.787617787518826	37.0	37.0	37.0	32.0	37.0
>>END_MODULE
>>Per tile sequence quality	pass
#Tile	Base	Mean
11101	1	0.014892076641700669
11101	2	0.5392987472679067
11101	3	0.4010864755168484
11101	4	0.2713441615718324
11101	5	0.17637754306385744
11101	6	0.17852678378774556
11101	7	0.111637561719796
11101	8	0.05598706949071186
11101	9	0.10213007404072272
11101	10	0.07561097978606313
11101	11	0.012403039582743247
11101	12	0.09689453355486677
11101	13	0.025568448080051098
11101	14	0.08043838282198834
11101	15	0.06901000097775523
11101	16	0.09494920356743108
11101	17	0.10268398595377448
11101	18	0.04058607483872123
11101	19	0.09020216132553571
11101	20	0.13873435209394103
11101	21	0.061406510663829295
11101	22	0.14610322844060164
11101	23	-0.004694005429279002
11101	24	0.08478138316886685
11101	25	0.0683295689058383
11101	26	0.05385292274518605
11101	27	0.11247880935551535
11101	28	0.08342215271520814
11101	29	0.14691943742069213
11101	30	0.08793508426292362
11101	31	0.07672858969590379
11101	32	0.023717795510343365
11101	33	0.058189986675685645
11101	34	0.04077113088194295
11101	35	0.025546319223550995
11101	36	0.0011217314001328305
11101	37	0.12665531740773162
11101	38	0.05771661241046644
11101	39	0.08558208276034662
11101	40	0.10431152549907807
11101	41	-0.006305085230863483
11101	42	0.07150962572423936
11101	43	0.13298805617888831
11101	44	0.11684510832024131
11101	45	0.14622260334986237
11101	46	0.14254034737708565
11101	47	0.09419999883362351
11101	48	0.023404593370962345
11101	49	0.019739373254338943
11101	50	0.02015478640190338
11101	51	0.050838359330128924
11101	52	0.08596724194129024
11101	53	0.03966949795226071
11101	54	0.17556711091699384
11101	55	-0.08831697915216097
11101	56	-0.008152009657642623
11101	57	0.04072526309241198
11101	58	0.07167084315835126
11101	59	0.03606299241975108
11101	60	0.08510756783174145
11101	61	0.09410384812336758
11101	62	0.12846218240770213
11101	63	0.058471008698695925
11101	64	0.0422801338170018
11101	65	0.1191285298858702
11101	66	0.08491443020329115
11101	67	0.005837035529332013
11101	68	0.12261902342466158
11101	69	-0.0035759688251886246
11101	70	0.06207334190192171
11101	71	0.08452671161314385
11101	72	-0.03977224484759034
11101	73	-0.04048845081952379
11101	74	0.0752050163654161
11101	75	0.00794912287007321
11101	76	0.05878684577317017
11102	1	0.6339825030895412
11102	2	-0.5472502508799977
11102	3	-0.12003794908153509
11102	4	0.08164783906383377
11102	5	-0.07090717070790475
11102	6	0.009545653338406623
11102	7	0.05277506320733494
11102	8	0.020977124444293338
11102	9	0.004416521788968453
11102	10	0.20989261948494686
11102	11	-0.0019454910402458836
11102	12	0.001671778194321405
11102	13	-0.1482510256823062
11102	14	-0.05099809120157062
11102	15	0.04227699760191683
11102	16	-0.010143920098883541
11102	17	0.0620007352272367
11102	18	-0.026404537500084757
11102	19	-0.04755142356149378
11102	20	0.06150573438406326
11102	21	0.047801179914642944
11102	22	-0.12920092488953827
11102	23	-0.06956862046565959
11102	24	0.16446645288282014
11102	25	0.13427204574050933
11102	26	0.011004417690060109
11102	27	0.1597906820072552
11102	28	0.12749773894280025
11102	29	-0.14200865747780966
11102	30	0.1640232755924771
11102	31	0.11563265785071764
11102	32	-0.07849816596185377
11102	33	0.03402022735467369
11102	34	0.13091706612345178
11102	35	-0.028368402010741534
11102	36	0.013185124223703326
11102	37	-0.041513459573423006
11102	38	0.03715066062743233
11102	39	0.04472694593029303
11102	40	-0.03129594849459494
11102	41	-0.0412185859092844
11102	42	0.2512273360644741
11102	43	0.19962046526435984
11102	44	-0.04176596391166498
11102	45	-0.08215923214343945
11102	46	-0.057071520793478214
11102	47	-0.06106573584780506
11102	48	0.04989294656410692
11102	49	0.026020455585090474
11102	50	-0.09907189259406124
11102	51	-0.009357808200128659
11102	52	-0.04294334954514767
11102	53	0.01365332102390937
11102	54	0.15222108933971867
11102	55	-0.12667076692289925
11102	56	0.11579087763510643
11102	57	-0.034266397882539934
11102	58	0.04966143491481745
11102	59	0.05976261005069716
11102	60	0.021196769774931568
11102	61	0.12523719665607302
11102	62	-0.0383306001085586
11102	63	0.06044848162947858
11102	64	-0.2564464682392682
11102	65	0.004749616066582973
11102	66	-0.15892516913117305
11102	67	0.024140999384478334
11102	68	0.06927902122315999
11102	69	0.15992497695035013
11102	70	-0.03017221027612038
11102	71	0.023157891799591823
11102	72	-0.018819724529187454
11102	73	0.045570988913340216
11102	74	0.03519469118577945
11102	75	-0.12200860868816932
11102	76	0.08919783541625748
>>END_MODULE
>>Per sequence quality scores	pass
#Quality	Count
2	378.0
3	0.0
4	0.0
5	0.0
6	0.0
7	0.0
8	0.0
9	0.0
10	0.0
11	0.0
12	0.0
13	0.0
14	0.0
15	0.0
16	0.0
17	1.0
18	0.0
19	4.0
20	14.0
21	35.0
22	109.0
23	208.0
24	393.0
25	719.0
26	1244.0
27	1957.0
28	2900.0
29	3963.0
30	5418.0
31	7163.0
32	9435.0
33	13175.0
34	18958.0
35	33914.0
36	235047.0
37	44953.0
>>END_MODULE
>>Per base sequence content	fail
#Base	G	A	T	C
1	61.87297489528727	8.874107636785121	4.119491056610732	25.13342641131688
2	31.41159889885539	9.43875870971141	30.063882193332542	29.085760198100658
3	27.028636160067443	15.680866197739665	22.199214942437894	35.091282699755
4	34.08864581192798	20.47905832266359	19.68212067085026	25.750175194558167
5	32.21967854875379	26.41448295665598	20.89722885948297	20.46860963510726
6	29.51312657970961	27.172981990980993	21.043642813655545	22.270248615653852
7	23.75464292299992	19.70232607149443	35.33494902663259	21.208081978873054
8	25.553808097492354	18.340406586950966	27.385741269076053	28.72004404648063
9	28.511893785727455	16.998235030689393	28.91968072495456	25.570190458628595
10	29.827741087399207	28.555622584647644	21.31516347341828	20.301472854534868
11	31.452016543294437	20.49234740918311	18.83670082453044	29.21893522299202
12	31.525671121930916	17.663238932317395	22.332967111139215	28.478122834612474
13	29.136482179078527	20.45968230552409	23.71354811517083	26.69028740022655
14	29.43758067490319	20.839545849687838	23.0536603356076	26.66921313980137
15	31.208872263638998	19.896472695661334	22.497826716893655	26.39682832380601
16	30.11942784260401	20.41683795549653	22.908695148046256	26.555039053853207
17	29.787362968856616	20.996293144556823	22.58468201924825	26.63166186733831
18	29.072175814909485	21.1381064059659	21.971329942817995	27.818387836306623
19	30.035826242722795	20.739969969178894	22.771792102420907	26.452411685677408
20	30.317510912544222	20.939703747803662	21.930460025974092	26.81232531367802
21	30.86193725138958	20.139090118806145	22.403255973235687	26.595716656568584
22	30.631326037588718	21.942821313971685	21.223345680248272	26.202506968191326
23	30.153637024626185	20.503640712757772	21.84190560490627	27.500816657709777
24	30.840316432521004	20.30415666729012	22.45638252192399	26.39914437826489
25	31.350565325971274	20.614903952540015	20.995563798063245	27.038966923425466
26	30.572322089582908	20.597061326434098	21.641958167241455	27.188658416741536
27	31.357765150018306	21.238345886450276	21.88957999088483	25.51430897264659
28	30.062514983310894	20.677627644636697	21.506418784527476	27.753438587524926
29	28.851980461484146	21.29529610757776	22.907697901243022	26.945025529695066
30	30.874897232118386	20.883487572991548	21.150157050403692	27.091458144486367
31	30.231075277316684	21.51081600927463	21.32374252364767	26.934366189761022
32	29.939766337656316	20.523184426726253	22.775754765204653	26.761294470412782
33	29.672672459001703	21.154408683759062	21.07535042572872	28.09756843151051
34	33.10249781944763	22.077466922799225	20.08716939268652	24.73286586506663
35	30.38230542796229	21.75322365144244	21.684178347708986	26.180292572886277
36	30.25695084991435	20.828304124390566	21.888259322703913	27.02648570299117
37	32.02895761547851	20.503383749041028	20.238167423737124	27.229491211743333
38	31.079977018401095	21.289868593747528	20.557461903783004	27.072692484068376
39	30.36441942234804	20.840383976721018	20.720984296174468	28.074212304756486
40	31.0489982366012	20.62939367021543	20.363170533052525	27.958437560130843
41	31.77918736224915	20.4155989327934	20.78996488341928	27.01524882153817
42	29.580223610082225	21.81740531106899	21.51080483284606	27.091566246002724
43	32.9064223860604	20.347751795715595	20.569249754770116	26.176576063453894
44	30.689570603918177	20.504368326534056	20.59666620078216	28.209394868765607
45	30.400601297536788	21.314415317263567	21.265362097157023	27.019621288042618
46	29.592531054672044	20.46443547749031	21.013793285333758	28.92924018250389
47	29.28254906098333	19.35218400506436	21.656995146655415	29.708271787296898
48	29.307051884054914	19.773871202608447	22.71470628581076	28.204370627525876
49	30.017792466975703	20.37237468717992	21.534164704391625	28.075668141452752
50	30.282255085130416	20.105924823458626	22.096699283278973	27.51512080813199
51	30.022122842494863	19.932469891286555	22.424061627164104	27.621345639054475
52	29.58984787661103	21.42985421508557	21.328174519332347	27.652123388971056
53	29.92978952706986	20.71631587844977	22.63323612695998	26.72065846752039
54	29.693553416736695	20.383613194433433	22.111933869271304	27.810899519558564
55	30.312345538694952	21.227708012296112	20.188670767132486	28.27127568187645
56	29.727686125211505	20.524270304568528	21.229378172588834	28.518665397631132
57	29.563035032832534	19.568350650964877	22.379545712993647	28.48906860320894
58	30.430110941734416	21.233485772357724	20.725355691056908	27.611047594850945
59	30.35784246248981	20.27376880803888	20.81431793394818	28.554070795523128
60	29.61741686861865	20.2455206388304	21.639187954391957	28.49787453815899
61	29.986851592653856	19.809560164567163	22.42386648004411	27.779721762734873
62	30.58352715285531	20.524625466426567	22.461141064367528	26.430706316350587
63	29.290699217195172	20.38742205121401	22.36513201539074	27.956746716200083
64	30.494556583372738	18.88299709365652	22.411521234372426	28.210925088598316
65	31.36868686868687	19.37612971823498	22.24375332270069	27.01143009037746
66	29.61867857542063	20.997961388810527	21.417659802100356	27.96570023366849
67	30.01537104968019	20.230912060013694	21.668384920920346	28.08533196938577
68	28.8128544121058	19.129655845350353	22.000943765946047	30.056545976597803
69	29.31796893138592	19.263768317058915	21.629007336163898	29.78925541539127
70	30.417288944340992	20.093401763839946	21.35576828026941	28.13354101154965
71	31.146821520799605	19.783507499632044	21.708123151852497	27.36154782771586
72	30.68803794887205	19.67685505162163	21.546395071798063	28.08871192770826
73	32.02901558873557	19.8464578284673	21.62960646731703	26.494920115480102
74	31.009323397130757	20.27127583905103	21.658295306566867	27.06110545725135
75	31.340604921949293	16.26481571683715	21.631236517987613	30.763342843225942
76	32.46262502611392	0.0	28.8403953922219	38.69697958166418
>>END_MODULE
>>Per sequence GC content	fail
#GC Content	Count
0	383.0
1	194.0
2	2.5
3	2.0
4	4.0
5	5.5
6	10.5
7	31.0
8	48.0
9	27.5
10	9.0
11	11.0
12	11.0
13	27.5
14	66.0
15	97.5
16	112.5
17	97.5
18	194.0
19	405.0
20	487.5
21	386.5
22	308.0
23	386.5
24	441.0
25	463.0
26	619.0
27	781.0
28	933.5
29	1116.0
30	1723.0
31	2366.0
32	2576.0
33	2833.0
34	3847.5
35	4400.5
36	3855.0
37	3390.0
38	3451.5
39	3811.5
40	4494.0
41	5475.5
42	6020.5
43	7011.0
44	8165.0
45	10546.0
46	12945.5
47	15506.0
48	15858.0
49	14363.5
50	14618.5
51	16301.5
52	17653.5
53	17592.0
54	17348.5
55	15777.5
56	15497.0
57	16449.0
58	16966.0
59	16200.5
60	14229.5
61	12520.5
62	11457.5
63	10760.5
64	10181.5
65	8558.5
66	6979.5
67	6524.5
68	7462.5
69	7558.0
70	5574.0
71	4362.0
72	4217.5
73	5378.0
74	5561.0
75	5156.0
76	5686.5
77	5401.5
78	4605.5
79	4628.5
80	5793.5
81	5343.0
82	4179.5
83	4405.0
84	6078.5
85	6786.5
86	3908.5
87	1902.5
88	2359.0
89	3164.5
90	2408.5
91	905.0
92	536.5
93	489.5
94	352.5
95	226.0
96	190.0
97	176.0
98	137.0
99	69.0
100	28.0
>>END_MODULE
>>Per base N content	pass
#Base	N-Count
1	0.09947682558396581
2	0.10079265661020874
3	0.10474014968893755
4	0.10789814415192059
5	0.10447698348368896
6	0.14921523837594872
7	0.09947682558396581
8	0.10237165384170026
9	0.09947682558396581
10	0.10131898902070592
11	0.09947682558396581
12	0.11500363169363242
13	0.09947682558396581
14	0.09947682558396581
15	0.09947682558396581
16	0.13474109708727644
17	0.11079297240965504
18	0.13105677021379622
19	0.09947682558396581
20	0.0997399917892144
21	0.09947682558396581
22	0.10684547933092624
23	0.10316115245744603
24	0.10026632419971157
25	0.10105582281545733
26	0.11289830205164374
27	0.10552964830468331
28	0.10500331589418613
29	0.11316146825689233
30	0.1294777729823047
31	0.12053012200385275
32	0.12263545164584144
33	0.13710959293451372
34	0.13079360400854764
35	0.13868859016600524
36	0.0036894145426180085
37	0.03689521495415769
38	0.0
39	0.003426363494996191
40	0.003690085530911056
41	0.01502510774583844
42	0.0018453686519318375
43	0.01318266426917946
44	0.012392749977587578
45	0.010020621384006034
46	0.0
47	0.0021101052942541832
48	0.0013189793210422046
49	0.05672419115260918
50	0.001319424836325349
51	0.01979576053992277
52	0.03458554113171283
53	0.022183489145935667
54	0.03698400705866192
55	0.02061261756459281
56	0.008723858165927783
57	0.0
58	0.028309649357212214
59	0.004235269204298798
60	0.006620938054503562
61	0.04372053831906454
62	0.027043436531440646
63	0.011409073662816767
64	0.04567301309116015
65	0.013289036544850499
66	0.048147645686529565
67	0.05165413033348865
68	0.02532063925285457
69	0.014141931627763013
70	0.007213002639424669
71	0.01926385325224678
72	0.0029512533436358906
73	2.695606700200014E-4
74	0.028443588705707318
75	0.006957973838018369
76	0.0
>>END_MODULE
>>Sequence Length Distribution	warn
#Length	Count
35	524.0
36	11.0
37	19.0
38	23.0
39	16.0
40	30.0
41	37.0
42	42.0
43	32.0
44	36.0
45	48.0
46	42.0
47	47.0
48	54.0
49	74.0
50	84.0
51	98.0
52	111.0
53	118.0
54	133.0
55	136.0
56	142.0
57	168.0
58	183.0
59	190.0
60	193.0
61	226.0
62	278.0
63	303.0
64	340.0
65	323.0
66	352.0
67	387.0
68	416.0
69	448.0
70	567.0
71	1034.0
72	1749.0
73	5338.0
74	20708.0
75	72085.0
76	272843.0
>>END_MODULE
>>Sequence Duplication Levels	fail
#Total Deduplicated Percentage	39.0958974034817
#Duplication Level	Percentage of total
1	34.413393047069995
2	4.2292688859395104
3	1.9722575226717567
4	1.2680464075212772
5	1.0428983513617631
6	0.9202475480158026
7	0.717802189536058
8	0.7082105897502663
9	0.6733068805951249
>10	15.273708710399417
>50	9.697387969765732
>100	22.353220725266553
>500	4.946018179000487
>1k	1.784232993106257
>5k	0.0
>10k+	0.0
>>END_MODULE
>>Overrepresented sequences	warn
#Sequence	Count	Percentage	Possible Source
GGCGTACGGAAGACCCGCTCCCCGGCGCCGCTCGTGGGGGGCCCAAGTCC	2067	0.5439645462488288	No Hit
GTACGGAAGACCCGCTCCCCGGCGCCGCTCGTGGGGGGCCCAAGTCCTTC	1273	0.335010579281451	No Hit
GTTGGATTGTTCACCCACTAATAGGGAACGTGAGCTGGGTTTAGACCGTC	1201	0.3160626125035528	No Hit
GGGAGTTTGACTGGGGCGGTACACCTGTCAAACGGTAACGCAGGTGTCCT	1158	0.3047464656778635	No Hit
GCCGAAGTGGAGAAGGCTTCCATGTGAACAGCAGTTGAACATGGGTCAGT	1075	0.2829036706422308	No Hit
CTCTGATCGTTTTTTCACTGACCCGGTGAGGCGGGGGGGCGAGCCCCGAG	864	0.227375601334779	No Hit
GTTTGGGAGCCGCGGAGGCGGCCGCGCCGAGCCGGGCCCGTGGCCCGCCG	831	0.21869111656157564	No Hit
GGCGGAGATGGGCGCCGCGAGGCGTCCAGTGCGGTAACGCGACCGATCCC	824	0.21684895312483551	No Hit
GCCCTTCGTCCTGGGAAACGGGGCGCGGCCGGAAAGGCGGCCGCCCCCTC	798	0.21000663178837226	No Hit
GCCGAAACGATCTCAACCTATTCTCAAACTTTAAATGGGTAAGAAGCCCG	770	0.20263797804141184	No Hit
GTTTTAAGCAGGAGGTGTCAGAAAAGTTACCACAGGGATAACTGGCTTGT	763	0.20079581460467175	No Hit
GTCCTTTCGTACTAAAATATCATAATTTTTTAAAGATAGAAACCAACCTG	760	0.20000631598892596	No Hit
GGTCAGTCGGTCCTGAGAGATGGGCGAGCGCCGTTCCGAAGGGACGGGCG	757	0.19921681737318023	No Hit
GCTGGGTTTAGACCGTCGTGAGACAGGTTAGTTTTACCCTACTGATGATG	746	0.19632198911544574	No Hit
GGGTAAACGGCGGGAGTAACTATGACTCTCTTAAGGTAGCCAAATGCCTC	735	0.1934271608577113	No Hit
GTCGGCTCTTCCTATCATTGTGAAGCAGAATTCACCAAGCGTTGGATTGT	710	0.18684800572649662	No Hit
GGCAGGGCGAAGCCAGAGGAAACTCTGGTGGAGGTCCGTAGCGGTCCTGA	710	0.18684800572649662	No Hit
GCGGAGATGGGCGCCGCGAGGCGTCCAGTGCGGTAACGCGACCGATCCCG	710	0.18684800572649662	No Hit
GCTTGATCTTGATTTTCAGTACGAATACAGACCGTGAAAGCGGGGCCTCA	681	0.1792161857742876	No Hit
GTCCGCTCCGGGCCGGGACGGGGTCCGGGGAGCGTGGTTTGGGAGCCGCG	672	0.17684768992705033	No Hit
CCCGTCCCGGGGGTTGGCCGCGCGGGCCCCGGTGGGGCGGCCACCCGGGG	616	0.16211038243312947	No Hit
GGGATCCCGAGGCCTCTCCAGTCCGCCGAGGGCGCACCACCGGCCCGTCT	607	0.1597418865858922	No Hit
GGGGAGTTTGACTGGGGCGGTACACCTGTCAAACGGTAACGCAGGTGTCC	597	0.1571102245334063	No Hit
GTCCCCGTCCCGGGGGTTGGCCGCGCGGGCCCCGGTGGGGCGGCCACCCG	596	0.15684705832815773	No Hit
GGTGTATGTGCTTGGCTGAGGAGCCAATGGGGCGAAGCTACCATCTGTGG	596	0.15684705832815773	No Hit
GTTGTTGCCATGGTAATCCTGCTCAGTACGAGAGGAACCGCAGGTTCAGA	578	0.15211006663368318	No Hit
GCCCAGTGCTCTGAATGTCAAAGTGAAGAAATTCAATGAAGCGCGGGTAA	576	0.151583734223186	No Hit
GTCGCGGTTCCGGCGGCGTCCGGTGAGCTCTCGCTGGCCCTTGAAAATCC	574	0.1510574018126888	No Hit
GTCGACCCGTGCGGAGGAGCGAGGAGGAAGGACGCGCGAGGGCCGGGACC	567	0.14921523837594872	No Hit
GTGGCGGCCAAGCGTTCATAGCGACGTCGCTTTTTGATCCTTCGATGTCG	563	0.14816257355495438	No Hit
GGCGCAATGAAGGTGAAGGCCGGCGCGCTCGCCGGCCGAGGTGGGATCCC	531	0.1397412549869996	No Hit
GTCGCTTTTTGATCCTTCGATGTCGGCTCTTCCTATCATTGTGAAGCAGA	524	0.13789909155025948	No Hit
GGCACGGTGAAGAGACATGAGAGGTGTAGAATAAGTGGGAGGCCCCCGGC	522	0.1373727591397623	No Hit
GGATTGTTCACCCACTAATAGGGAACGTGAGCTGGGTTTAGACCGTCGTG	490	0.12895144057180752	No Hit
GGCCGATCGAAAGGGAGTCGGGTTCAGATCCCCGAATCCGGAGTGGCGGA	484	0.127372443340316	No Hit
GTCGTTTACAATTGATTCTAGGCTTTGTCATTGTATTAAATAATGCTTTT	474	0.12474078128783014	No Hit
GCGTACGGAAGACCCGCTCCCCGGCGCCGCTCGTGGGGGGCCCAAGTCCT	459	0.12079328820910135	No Hit
CCGGGGGTTGGCCGCGCGGGCCCCGGTGGGGCGGCCACCCGGGGTCCCGG	451	0.11868795856711264	No Hit
GGGCGGTACACCTGTCAAACGGTAACGCAGGTGTCCTAAGGCGAGCTCAG	446	0.11737212754086972	No Hit
GGTAAAGCGAATGATTAGAGGTCTTGGGGCCGAAACGATCTCAACCTATT	445	0.11710896133562113	No Hit
GTGCGGAGTGCCCTTCGTCCTGGGAAACGGGGCGCGGCCGGAAAGGCGGC	445	0.11710896133562113	No Hit
GGTCGACCCGTGCGGAGGAGCGAGGAGGAAGGACGCGCGAGGGCCGGGAC	441	0.11605629651462677	No Hit
GTCAGTCGGTCCTGAGAGATGGGCGAGCGCCGTTCCGAAGGGACGGGCGA	439	0.1155299641041296	No Hit
GTAGAATAAGTGGGAGGCCCCCGGCGCCCCCCCGGTGTCCCCGCGAGGGG	437	0.11500363169363242	No Hit
GCCTAGTGGGCCACTTTTGGTAAGCAGAACTGGCGCTGCGGGATGAACCG	426	0.11210880343589798	No Hit
GTTTTTTCACTGACCCGGTGAGGCGGGGGGGCGAGCCCCGAGGGGCTCTC	424	0.11158247102540081	No Hit
>>END_MODULE
>>Adapter Content	pass
#Position	Illumina Universal Adapter	Illumina Small RNA 3' Adapter	Illumina Small RNA 5' Adapter	Nextera Transposase Sequence	PolyA	PolyG
1	0.0	0.0	0.0	0.0	0.0010526648209943472	0.0
2	0.0	0.0	2.631662052485868E-4	0.0	0.0021053296419886944	0.0
3	0.0	0.0	2.631662052485868E-4	0.0	0.0021053296419886944	5.263324104971736E-4
4	0.0	0.0	2.631662052485868E-4	0.0	0.0028948282577344548	7.894986157457604E-4
5	0.0	0.0	2.631662052485868E-4	0.0	0.0031579944629830416	7.894986157457604E-4
6	0.0	0.0	2.631662052485868E-4	0.0	0.0031579944629830416	0.0010526648209943472
7	0.0	0.0	5.263324104971736E-4	0.0	0.004210659283977389	0.0010526648209943472
8	0.0	0.0	5.263324104971736E-4	0.0	0.0057896565154689095	0.001315831026242934
9	0.0	0.0	5.263324104971736E-4	0.0	0.00736865374696043	0.001315831026242934
10	0.0	0.0	5.263324104971736E-4	0.0	0.008421318567954778	0.001315831026242934
11	0.0	0.0	5.263324104971736E-4	0.0	0.008684484773203365	0.001315831026242934
12	0.0	0.0	5.263324104971736E-4	0.0	0.008947650978451951	0.0015789972314915208
13	0.0	0.0	5.263324104971736E-4	0.0	0.008947650978451951	0.0015789972314915208
14	0.0	0.0	5.263324104971736E-4	0.0	0.009473983388949125	0.0015789972314915208
15	0.0	0.0	5.263324104971736E-4	0.0	0.009737149594197712	0.0015789972314915208
16	0.0	0.0	5.263324104971736E-4	0.0	0.010000315799446298	0.0015789972314915208
17	0.0	0.0	5.263324104971736E-4	0.0	0.010263482004694886	0.0015789972314915208
18	0.0	0.0	5.263324104971736E-4	0.0	0.011316146825689233	0.0015789972314915208
19	0.0	0.0	5.263324104971736E-4	0.0	0.011579313030937819	0.0015789972314915208
20	0.0	0.0	5.263324104971736E-4	0.0	0.011579313030937819	0.0015789972314915208
21	0.0	0.0	5.263324104971736E-4	0.0	0.012105645441434993	0.0015789972314915208
22	0.0	0.0	5.263324104971736E-4	0.0	0.012631977851932166	0.0015789972314915208
23	0.0	0.0	5.263324104971736E-4	0.0	0.012895144057180754	0.0015789972314915208
24	0.0	0.0	5.263324104971736E-4	0.0	0.01315831026242934	0.0018421634367401076
25	0.0	0.0	5.263324104971736E-4	0.0	0.013421476467677928	0.0018421634367401076
26	0.0	0.0	5.263324104971736E-4	0.0	0.013421476467677928	0.0021053296419886944
27	0.0	0.0	5.263324104971736E-4	0.0	0.013947808878175101	0.0021053296419886944
28	0.0	0.0	5.263324104971736E-4	0.0	0.01473730749392086	0.002368495847237281
23	0.0	0.0	5.263324104971736E-4	0.0	0.012895144057180754	0.0015789972314915208
24	0.0	0.0	5.263324104971736E-4	0.0	0.01315831026242934	0.0018421634367401076
25	0.0	0.0	5.263324104971736E-4	0.0	0.013421476467677928	0.0018421634367401076
26	0.0	0.0	5.263324104971736E-4	0.0	0.013421476467677928	0.0021053296419886944
27	0.0	0.0	5.263324104971736E-4	0.0	0.013947808878175101	0.0021053296419886944
28	0.0	0.0	5.263324104971736E-4	0.0	0.01473730749392086	0.002368495847237281
29	0.0	0.0	5.263324104971736E-4	0.0	0.015000473699169448	0.004210659283977389
30	0.0	0.0	5.263324104971736E-4	0.0	0.015263639904418034	0.0057896565154689095
31	0.0	0.0	5.263324104971736E-4	0.0	0.015789972314915208	0.006315988925966083
32	0.0	0.0	5.263324104971736E-4	0.0	0.016579470930660967	0.008421318567954778
33	0.0	0.0	5.263324104971736E-4	0.0	0.01736896954640673	0.009473983388949125
34	0.0	0.0	5.263324104971736E-4	0.0	0.019211132983146837	0.010000315799446298
35	0.0	0.0	5.263324104971736E-4	0.0	0.020790130214638356	0.011842479236186407
36	0.0	0.0	5.263324104971736E-4	0.0	0.02131646262513553	0.013421476467677928
37	0.0	0.0	5.263324104971736E-4	0.0	0.02210596124088129	0.015000473699169448
38	0.0	0.0	5.263324104971736E-4	0.0	0.02236912744612988	0.016842637135909555
39	0.0	0.0	5.263324104971736E-4	0.0	0.022632293651378466	0.020263797804141184
40	0.0	0.0	5.263324104971736E-4	0.0	0.022632293651378466	0.023158626061875638
41	0.0	0.0	5.263324104971736E-4	0.0	0.023948124677621398	0.026579786730107267
42	0.0	0.0	5.263324104971736E-4	0.0	0.023948124677621398	0.027895617756350202
43	0.0	0.0	5.263324104971736E-4	0.0	0.023948124677621398	0.030000947398338897
44	0.0	0.0	7.894986157457604E-4	0.0	0.023948124677621398	0.03210627704032759
45	0.0	0.0	7.894986157457604E-4	0.0	0.024474457088118573	0.033948440477067694
46	0.0	0.0	7.894986157457604E-4	0.0	0.025000789498615745	0.03552743770855922
47	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.025263955703864333	0.03815909976104508
48	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.025263955703864333	0.04000126319778519
49	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.02552712190911292	0.04289609145551965
50	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.025790288114361508	0.0457909197132541
51	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.025790288114361508	0.047106750739497036
52	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.026053454319610092	0.05000157899723149
53	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.026053454319610092	0.05342273966546312
54	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.026053454319610092	0.055791235512700404
55	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.026053454319610092	0.0597387285914292
56	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.026579786730107267	0.06605471751739529
57	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.026842952935355855	0.06947587818562692
58	2.631662052485868E-4	0.0	7.894986157457604E-4	0.0	0.027369285345853027	0.07210754023811278
59	5.263324104971736E-4	0.0	7.894986157457604E-4	0.0	0.027895617756350202	0.07605503331684159
60	5.263324104971736E-4	0.0	0.0010526648209943472	0.0	0.027895617756350202	0.08052885880606755
61	5.263324104971736E-4	0.0	0.0010526648209943472	0.0	0.028421950166847374	0.08316052085855342
62	5.263324104971736E-4	0.0	0.0010526648209943472	0.0	0.028685116372095962	0.08816067875827657
63	5.263324104971736E-4	0.0	0.0010526648209943472	0.0	0.029211448782593134	0.09421350147899407
64	5.263324104971736E-4	0.0	0.0010526648209943472	0.0	0.029211448782593134	0.10105582281545733
65	5.263324104971736E-4	0.0	0.0010526648209943472	0.0	0.029211448782593134	0.10684547933092624
>>END_MODULE
"""
}

FASTQ_SCREEN_V0_9_2 = {
    'screen.txt': """#Fastq_screen version: 0.9.2	#Aligner: bowtie	#Reads in subset: 1000000
Genome	#Reads_processed	#Unmapped	%Unmapped	#One_hit_one_genome	%One_hit_one_genome	#Multiple_hits_one_genome	%Multiple_hits_one_genome	#One_hit_multiple_genomes	%One_hit_multiple_genomes	Multiple_hits_multiple_genomes	%Multiple_hits_multiple_genomes
hg19	203860	202199	99.18	2	0.00	0	0.00	136	0.07	1523	0.75
mm9	203860	201733	98.96	89	0.04	62	0.03	58	0.03	1918	0.94
rn4	203860	201822	99.00	48	0.02	21	0.01	93	0.05	1876	0.92
dm3	203860	203844	99.99	0	0.00	0	0.00	0	0.00	16	0.01
ws200	203860	203853	100.00	0	0.00	0	0.00	1	0.00	6	0.00
ecoli	203860	203860	100.00	0	0.00	0	0.00	0	0.00	0	0.00
saccer	203860	203521	99.83	0	0.00	1	0.00	79	0.04	259	0.13
PhiX	203860	203860	100.00	0	0.00	0	0.00	0	0.00	0	0.00
Vectors	203860	203860	100.00	0	0.00	0	0.00	0	0.00	0	0.00
SpR6	203860	203860	100.00	0	0.00	0	0.00	0	0.00	0	0.00

%Hit_no_genomes: 98.92
"""
}

FASTQ_SCREEN_V0_15_3 = {
    'screen.txt': """#Fastq_screen version: 0.15.3	#Aligner: bowtie	#Reads in subset: 100000
Genome	#Reads_processed	#Unmapped	%Unmapped	#One_hit_one_genome	%One_hit_one_genome	#Multiple_hits_one_genome	%Multiple_hits_one_genome	#One_hit_multiple_genomes	%One_hit_multiple_genomes	Multiple_hits_multiple_genomes	%Multiple_hits_multiple_genomes
hg19	94997	43983	46.30	6630	6.98	14774	15.55	3648	3.84	25962	27.33
mm9	94997	72416	76.24	4	0.00	1	0.00	5767	6.07	16809	17.69
rn4	94997	65932	69.41	1	0.00	0	0.00	2501	2.63	26563	27.96
dm3	94997	55447	58.37	26266	27.65	11117	11.70	95	0.10	2072	2.18
ws200	94997	93187	98.09	0	0.00	0	0.00	634	0.67	1176	1.24
ecoli	94997	94993	100.00	0	0.00	4	0.00	0	0.00	0	0.00
saccer	94997	92724	97.60	7	0.01	5	0.01	0	0.00	2261	2.38
PhiX	94997	94997	100.00	0	0.00	0	0.00	0	0.00	0	0.00
Vectors	94997	94990	100.00	1	0.00	4	0.00	2	0.00	0	0.00
SpR6	94997	94994	100.00	0	0.00	3	0.00	0	0.00	0	0.00

%Hit_no_genomes: 6.90
"""
}

FASTQ_STRAND_V0_0_4 = { 'fastq_strand.txt': """#fastq_strand version: 0.0.4	#Aligner: STAR	#Reads in subset: 10000
#Genome	1st forward	2nd reverse
human	5.17	103.76
"""
}

SEQ_LENS_JSON = """{
    "fastq": "%(fastq)s",
    "frac_reads_masked": 0.0,
    "frac_reads_padded": 0.2712291764912113,
    "max_length": 76,
    "mean_length": 75.31423877677945,
    "median_length": 76,
    "min_length": 65,
    "nreads": 37285443,
    "nreads_masked": 0,
    "nreads_padded": 101129,
    "seq_lengths_dist": {
        "65": 3,
        "66": 6,
        "67": 2,
        "68": 1,
        "69": 16,
        "70": 218,
        "71": 307,
        "72": 1289,
        "73": 1559577,
        "74": 5165990,
        "75": 10549970,
        "76": 20008064
    },
    "seq_lengths_masked_dist": {},
    "seq_lengths_padded_dist": {
        "71": 1,
        "72": 8,
        "73": 4088,
        "74": 13609,
        "75": 30989,
        "76": 52434
    }
}
"""

RSEQC_GENEBODY_COVERAGE_TXT = """Percentile	1	2	3	4	5	6	7	8	9	10
SMP1_S1_R1_001	3264.0	6663.0	10749.0	15281.0	19831.0	23679.0	26241.0	28684.0	31556.0	34055.0
SMP2_S2_R1_001	3071.0	7003.0	11984.0	17241.0	22453.0	26756.0	30356.0	33396.0	36437.0	39849.0
"""

PICARD_COLLECT_INSERT_SIZE_METRICS = """"## htsjdk.samtools.metrics.StringHeader
# CollectInsertSizeMetrics --Histogram_FILE Sample.insert_size_histogram.pdf --INPUT Sample.bam --OUTPUT Sample.insert_size_metrics.txt --DEVIATIONS 10.0 --MINIMUM_PCT 0.05 --METRIC_ACCUMULATION_LEVEL ALL_READS --INCLUDE_DUPLICATES false --ASSUME_SORTED true --STOP_AFTER 0 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 5 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false
## htsjdk.samtools.metrics.StringHeader
# Started on: Mon Oct 04 12:53:24 BST 2021

## METRICS CLASS	picard.analysis.InsertSizeMetrics
MEDIAN_INSERT_SIZE	MODE_INSERT_SIZE	MEDIAN_ABSOLUTE_DEVIATION	MIN_INSERT_SIZE	MAX_INSERT_SIZE	MEAN_INSERT_SIZE	STANDARD_DEVIATION	READ_PAIRS	PAIR_ORIENTATION	WIDTH_OF_10_PERCENT	WIDTH_OF_20_PERCENT	WIDTH_OF_30_PERCENT	WIDTH_OF_40_PERCENT	WIDTH_OF_50_PERCENT	WIDTH_OF_60_PERCENT	WIDTH_OF_70_PERCENT	WIDTH_OF_80_PERCENT	WIDTH_OF_90_PERCENT	WIDTH_OF_95_PERCENT	WIDTH_OF_99_PERCENT	SAMPLE	LIBRARY	READ_GROUP
139	103	37	28	753074	153.754829	69.675347	175847	FR	15	31	47	61	75	93	115	157	323	975	31983			

## HISTOGRAM	java.lang.Integer
insert_size	All_Reads.fr_count
28	1
29	1
30	1
31	1
32	1
33	2
34	3
"""

QUALIMAP_RNASEQ_RESULTS = """RNA-Seq QC report
-----------------------------------

>>>>>>> Input

    bam file = /path/to/SMP1_S1_R1_001.bam
    gff file = /path/to/annotation.gtf
    counting algorithm = uniquely-mapped-reads
    protocol = strand-specific-reverse


>>>>>>> Reads alignment

    reads aligned (left/right) = 175,847 / 175,847
    read pairs aligned  = 175,847
    total alignments = 728,678
    secondary alignments = 376,984
    non-unique alignments = 357,582
    aligned to genes  = 63,784
    ambiguous alignments = 160,764
    no feature assigned = 19,130
    not aligned = 0


>>>>>>> Reads genomic origin

    exonic =  63,784 (76.93%)
    intronic = 4,778 (5.76%)
    intergenic = 14,352 (17.31%)
    overlapping exon = 6,523 (7.87%)
    rRNA = 0 (0%)


>>>>>>> Transcript coverage profile

    5' bias = 0.73
    3' bias = 0.41
    5'-3' bias = 1.52


>>>>>>> Junction analysis

    reads at junctions = 28,238

    AGAT : 3.22%
    AGGT : 3.18%
    ATCT : 2.79%
    ACCT : 2.79%
    AGCT : 2.66%
    GCCT : 2.02%
    AGGA : 1.94%
    TTCT : 1.82%
    AGAA : 1.78%
    TCCT : 1.66%
    TGCT : 1.34%
"""

QUALIMAP_RAW_COVERAGE_PROFILE = \
"""#Transcript position	Transcript coverage profile
0.0	22.648923862529898
1.0	36.072848086812215
2.0	49.90661635235814
3.0	61.26286478838015
4.0	67.64863079300177
5.0	71.3660047159799
6.0	71.9260591904655
7.0	79.80361428985356
8.0	82.33718275058882
9.0	81.62987377703458
10.0	81.9124541146298
"""

# Base64 encoded generic example PNG
BASE64_PNG_DATA = "iVBORw0KGgoAAAANSUhEUgAAAEwAAAAqCAIAAACWWCo7AAABFElEQVR4nO2asW3DMBBFvwwv4YZ1BjDgwn1qNlqDN8vnGirEOjO4j2sVUp0JmIJQQwOOAyOxfbwH4YMQVOjdB0EV6khCPSRzzuM8ak2Sm0cP+T8wSTXUexLQd20r50i63k3DJBJyRlqS3/mSMcYQQlrSNEyud+XOQ4r5Fcej1JIiAQIA45y6zgMewJpBZF1LeTz8/Uteg7xpyrUkGUuTgHv+JkV+nvLpJNsgAhEPAPBAfKlvg1ua3O8zql36DMf3nckLNmTMGfNnKgng8IWXzkuaOCdNUgv1EaKA/t3t3vxyTiWHj6mJJk1SCyaphVXyrDmbalI1JqkFk9SCSWrBJLVgklowSS2YpBZMUgtNSHYt/Kz0Dck+c8rgvriAAAAAAElFTkSuQmCC"
